compiler_gym/datasets/benchmark.py

```diff
@@ -0,0 +1,350 @@
+
+        :param validation_callback: A callback that accepts a single
+            :class:`CompilerEnv <compiler_gym.envs.CompilerEnv>` argument and
+            returns an iterable sequence of zero or more :class:`ValidationError
+            <compiler_gym.ValidationError>` tuples. Validation callbacks must be
+            thread safe and must not modify the environment.
+        """
+        self._validation_callbacks.append(validation_callback)
+
+    def write_sources_to_directory(self, directory: Path) -> int:
+        """Write the source files for this benchmark to the given directory.
+
+        This writes each of the :attr:`benchmark.sources
+        <compiler_gym.datasets.Benchmark.sources>` files to disk.
+
+        :param directory: The directory to write results to. If it does not
+            exist, it is created.
+
+        :return: The number of files written.
+        """
+        directory = Path(directory)
+        directory.mkdir(exist_ok=True, parents=True)
+        uniq_paths = set()
+        for filename, contents in self.sources:
+            path = directory / filename
+            uniq_paths.add(path)
+            path.parent.mkdir(exist_ok=True, parents=True)
+            with open(path, "wb") as f:
+                f.write(contents)
+
+        return len(uniq_paths)
+
+    @classmethod
+    def from_file(cls, uri: str, path: Path):
+        """Construct a benchmark from the path to a file.
+
+        :param uri: The URI of the benchmark.
+
+        :param path: A filesystem path.
+
+        :return: A :class:`Benchmark <compiler_gym.datasets.Benchmark>` instance.
+        """
+        return cls(
+            proto=BenchmarkProto(
+                uri=uri, program=File(uri=f"file:///{Path(path).absolute()}")
+            ),
+        )
+
+    @classmethod
+    def from_file_contents(cls, uri: str, data: bytes):
+        """Construct a benchmark from a raw data array.
+
+        :param uri: The URI of the benchmark.
+
+        :param data: An array of bytes that will be passed to the compiler
+            service.
+        """
+        return cls(proto=BenchmarkProto(uri=uri, program=File(contents=data)))
+
+    def __eq__(self, other: Union[str, "Benchmark"]):
+        if isinstance(other, Benchmark):
+            return self.uri == other.uri
+        else:
+            return self.uri == other
+
+    def __ne__(self, other: Union[str, "Benchmark"]):
+        return not self == other
+
+
+class BenchmarkInitError(OSError):
+    """Base class for errors raised if a benchmark fails to initialize."""
+
+
+class BenchmarkWithSource(Benchmark):
+    """A benchmark which has a single source file."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._src_name = None
+        self._src_path = None
+
+    @classmethod
+    def create(
+        cls, uri: str, bitcode_path: Path, src_name: str, src_path: Path
```

Comment:
This is very LLVM - if there are going to be multiple compilers, should all references things like bitcode be removed?
Consider moving this to LLVMBenchmark

Reply:
Good catch! This was initially LLVM-specific. I refactored it bust missed a couple of bitcode references. Fixed